### PR TITLE
Fix solver crash in periodic_box

### PR DESCRIPTION
**Pull request by tjhei**
The CG for the velocity block would crash for the longest time with
    AztecOO::Iterate error code -2: numerical breakdown
in tests/periodic_box.prm

Small permutations of the mesh refinement or solver settings would make
the error go away. Instead, we now add a constraint to remove the
nullspace from the linear system.

### CODE REVIEW

**Comment by ian-r-rose** [comment]
Fantastic, I was just thinking about this test.  Is there any benefit to thinking about which would be the best velocity node to constrain?  I see you have just chosen DOFs 0-2.

I wonder whether it would also be worthwhile to add a constraint for problems with a rotational nullspace.  We have not really seen the same solver convergence issues with those, but in theory they might be vulnerable to the same issue.  It would be a bit more complicated, as only the component of the velocity orthogonal to the radial direction should be constrained, so there is some geometric reasoning involved.

**Comment by tjhei** [comment]
> Fantastic, I was just thinking about this test. Is there any benefit to thinking about which would be the best velocity node to constrain? I see you have just chosen DOFs 0-2.

The main goal is to make the 0,0 block invertible. A smarter way to pick the DoF might be the some unknown on a periodic face, but this is a bit more tricky. I hope it doesn't matter too much.

> I wonder whether it would also be worthwhile to add a constraint for problems with a rotational nullspace.

More tricky, indeed. I guess we can just ignore it until we have a solver crash in practice. :-) I mean, I don't even know how I would test the fix without a broken test...

File: include/aspect/simulator.h [hunk 1]
```diff
@@ -839,6 +839,15 @@ namespace aspect
                                             LinearAlgebra::Vector &vec);
 
       /**
+       * Fill @p constraints with a zero Dirichlet constraints for the first
```
Comment (bangerth) [inline include/aspect/simulator.h]: grammar: "a ... constraintS"

File: include/aspect/simulator.h [hunk 2]
```diff
@@ -839,6 +839,15 @@ namespace aspect
...
+       * velocity unknown in the domain for each velocity component, which is
+       * later being processed for translational or linear momentum removal.
+       * While this is not the correct value for the velocity component, it ensures
+       * that the linear system is invertible.
+       */
+      void setup_nullspace_constraints(ConstraintMatrix &constraints);
```
Comment (bangerth) [inline include/aspect/simulator.h]: the function really only does translational nullspaces. do you want to rename the function maybe, or is the intent to deal with all possible null spaces but (for now) simply not do anything about rotational modes? In the latter case, maybe a sentence in the documentation, or a comment in the code would help to understand what is supposed to happen in those cases.

File: source/simulator/nullspace.cc [hunk 3]
```diff
@@ -165,6 +165,28 @@ namespace aspect
 
   }
 
+
+  template <int dim>
+  void Simulator<dim>::setup_nullspace_constraints(ConstraintMatrix &constraints)
+  {
+    unsigned int flags[] = {(NullspaceRemoval::linear_momentum_x
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: make them `const`.

File: tests/periodic_box_freesurface/screen-output [hunk 4]
```diff
@@ -40,48 +40,48 @@ Number of free surface degrees of freedom: 3246
      reference_cp:                                  1250
      reference_thermal_diffusivity:                 1.13939e-06
 
-     RMS, max velocity:  0.626 m/year, 1.23 m/year
+     RMS, max velocity:  0.646 m/year, 8.25 m/year
```
Comment (bangerth) [inline tests/periodic_box_freesurface/screen-output]: uh, that's a large difference in the max velocity. can you visually inspect whether everything still looks alright? the average did not change very much, so it looks like a singularity somewhere.

File: tests/periodic_box_freesurface/screen-output [hunk 5]
```diff
@@ -40,48 +40,48 @@ Number of free surface degrees of freedom: 3246
...
    Solving temperature system... 10 iterations.
    Rebuilding Stokes preconditioner...
-   Solving Stokes system... 30+7 iterations.
+   Solving Stokes system... 30+4 iterations.
 
    Postprocessing:
-     RMS, max velocity:  0.438 m/year, 1 m/year
-     Topography min/max: -464.4 m, 509.8 m, 
+     RMS, max velocity:  0.563 m/year, 8.18 m/year
+     Topography min/max: -167.1 m, 153 m,
```
Comment (bangerth) [inline tests/periodic_box_freesurface/screen-output]: these are similarly radical differences

**Comment by tjhei** [comment]
Good catch @bangerth. It turns out I managed to break that other test. I hope the new logic is correct now. I am not very happy with the somewhat fragile way of picking correct velocity indices. Is there a simple way to do ```Assert(this_dof_belongs_to_velocity_component_d(idx, d))``` ?

File: source/simulator/nullspace.cc [hunk 6]
```diff
@@ -165,6 +165,54 @@ namespace aspect
 
   }
 
+  namespace
+  {
+    inline MPI_Datatype mpi_type_id (const unsigned int *)
+    {
+      return MPI_UNSIGNED;
+    }
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: We already have something like this in namespace `dealii::Utilities::MPI`.

File: source/simulator/nullspace.cc [hunk 7]
```diff
@@ -165,6 +165,54 @@ namespace aspect
...
+          types::global_dof_index idx = d;
+          while (!constraints.can_store_line(idx) || constraints.is_constrained(idx))
+            {
+              idx += dim;
+            }
+
+          // Make a reduction to find the smallest index (processors with a larger candidate
+          // just happened to not be able to store that index)
+          types::global_dof_index global_idx;
+          MPI_Allreduce(&idx, &global_idx, 1, mpi_type_id(&global_idx), MPI_MIN, mpi_communicator);
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: `Utilities::MPI::min` should do, no? (Ah, bummer -- I see that we only have `max` but not `min` :-( )

**Comment by bangerth** [comment]
Yes, that's fragile. I think you need to come from the elements, not from the DoFs. How about something like
```
for (cell=...)
  if (cell->is_locally_owned())
    {
       vector<global_dof_index> local_dof_indices (...);
       cell->get_dof_indices (local_dof_indices);
       global_dof_index first_velocity_dof[dim]
       for (d=0..dim)
         first_velocity_dof[d]
         = local_dof_indices[cell->get_fe().system_to_component_index (...introspection...x-velocity-component,
                                                                                                      d);

      ...
      break;
   }
```

File: source/simulator/nullspace.cc (hunk 6 above)
Comment (tjhei) [inline source/simulator/nullspace.cc]: Yes, that is where I stole it from. It is in an ```internal``` namespace, though.

File: source/simulator/nullspace.cc (hunk 7 above)
Comment (tjhei) [inline source/simulator/nullspace.cc]: I will tackle that later.

Comment (tjhei) [inline source/simulator/nullspace.cc]: (I mean adding it to deal.II)

**Comment by tjhei** [comment]
Okay, I think I addressed your comments. Please review again.

(I will make a PR for putting MPI::min into deal.II later)

File: source/simulator/nullspace.cc [hunk 8]
```diff
@@ -165,6 +165,88 @@ namespace aspect
...
+      std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
+      typename DoFHandler<dim>::active_cell_iterator cell;
+      for (cell = dof_handler.begin_active(); n_left_to_find>0 && cell != dof_handler.end(); ++cell)
+        if (cell->is_locally_owned())
+          {
+            cell->get_dof_indices (local_dof_indices);
+
+            for (unsigned int i=0; i<finite_element.dofs_per_cell; ++i)
+              {
+                const unsigned int component = finite_element.system_to_component_index(i).first;
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: couldn't you use `component_to_system_index(introspection.component_indices.velocities[d], 0)` instead?

File: source/simulator/nullspace.cc [hunk 9]
```diff
@@ -165,6 +165,88 @@ namespace aspect
...
+                  continue; // already found one
+
+                const types::global_dof_index idx = local_dof_indices[i];
+
+                if (constraints.can_store_line(idx) && !constraints.is_constrained(idx))
+                  {
+                    vel_idx[velocity_component] = idx;
+                    --n_left_to_find;
+                  }
+              }
```
Comment (bangerth) [inline source/simulator/nullspace.cc]: i think at this point, `n_left_to_find==0` (and if not, then that's bad). So once you find the first cell that is locally owned, you may as well terminate the outer loop with a `break`.

**Comment by bangerth** [comment]
I had just a couple of minor notes. Feel free to address and/or ignore before merging yourself.

My only major question is whether you investigated the substantially different output? It looks like it is ok now in your most recent version. (Or is this what you refer to in your "good catch" comment?)

File: source/simulator/nullspace.cc (hunk 8 above)
Comment (tjhei) [inline source/simulator/nullspace.cc]: But the first DoF could already be constrained, so I would have to search through all DoFs anyway. Or do you mean I should loop over all velocity FE DoFs j with ```component_to_system_index(...,j)```?

File: source/simulator/nullspace.cc (hunk 9 above)
Comment (tjhei) [inline source/simulator/nullspace.cc]: You are right, it is unlikely that all DoFs in a cell are constrained.

**Comment by tjhei** [comment]
> Or is this what you refer to in your "good catch" comment?

Yes, I ended up overwriting a periodic constraint by blindly setting the first DoF to zero. This broke the computation (and a large outflow at that point).

File: source/simulator/nullspace.cc (hunk 8 above)
Comment (bangerth) [inline source/simulator/nullspace.cc]: Ah, excellent point. That's why you need the loop over all dofs (and, in fact, over more than possibly just the first locally owned cell -- maybe a comment to this effect would be a good idea).

The thing with system_to_component_index vs component_to_system index was that you try to identify only x-velocity dofs. You can do this with the first of these functions by looking at all dofs and finding out which of those are x-velocities. Or you can simply loop over all x-velocity dofs and then using the second of the two functions to find out the how-manyth dof they correspond to on the local cell.

As far as performance is concerned, it won't matter -- you do this at most once per time step, on at most a small number of cells. It was just a matter of making the code easy to understand, but I'm not feeling particular about it.

**Pull request merged**